### PR TITLE
Recovery: multi-thread restore skiplist

### PR DESCRIPTION
**Pull request by sherry-1001**
image

I add `debug_check` define to check skiplist connection correctness. 

please review @KVKIT @JiayuZzz @ZiyanShi

### CODE REVIEW

File: engine/kv_engine.cpp [hunk 1]
```diff
@@ -275,6 +276,8 @@ Status KVEngine::RestoreSkiplist(DLDataEntry *pmem_data_entry, DataEntry *) {
     skiplists_.push_back(std::make_shared<Skiplist>(
         (DLDataEntry *)pmem_data_entry, key, id, pmem_allocator_, hash_table_));
     skiplist = skiplists_.back().get();
+    ConcurrentRebuildSorted::entries_offsets_.insert(
```
Comment (KVKIT) [inline engine/kv_engine.cpp:279]: Need to check if opt_restore_sorted is true?

File: engine/kv_engine.cpp [hunk 2]
```diff
@@ -380,9 +383,19 @@ Status KVEngine::RestoreSortedRecord(DLDataEntry *pmem_data_entry,
         return Status::MemoryOverflow;
       }
       new_hash_offset = (uint64_t)dram_node;
+      if (configs_.opt_restore_sorted &&
+          thread_res_[write_thread.id]
+                      .visited_skiplist_ids_[dram_node->GetSkipListId()]++ %
+                  10000 ==
```
Comment (KVKIT) [inline engine/kv_engine.cpp:389]: Can add some explanation of the hard code 10000?

File: engine/kv_engine.cpp (hunk 1 above)
Comment (sherry-1001) [reply engine/kv_engine.cpp:279]: Yes，It needs

File: engine/kv_engine.cpp (hunk 2 above)
Comment (sherry-1001) [reply engine/kv_engine.cpp]: I added

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -380,9 +385,22 @@ Status KVEngine::RestoreSortedRecord(DLDataEntry *pmem_data_entry,
         return Status::MemoryOverflow;
       }
       new_hash_offset = (uint64_t)dram_node;
+// RESTORE_SKIPLIST_STRIDE: Select a data entry every 10000 into restored
+// skiplist map for multi-thread restoring large skiplist.
+#define RESTORE_SKIPLIST_STRIDE 10000
+      if (configs_.opt_restore_sorted &&
+          thread_res_[write_thread.id]
+                      .visited_skiplist_ids_[dram_node->GetSkipListId()]++ %
```
Comment (KVKIT) [inline engine/kv_engine.cpp:393]: Clear visited_skiplist_ids_ after doing restore?

File: engine/kv_engine.hpp [hunk 4]
```diff
@@ -68,6 +73,7 @@ class KVEngine : public Engine {
 
     alignas(64) uint64_t newest_restored_ts = 0;
     PendingBatch *persisted_pending_batch = nullptr;
+    std::unordered_map<uint64_t, int> visited_skiplist_ids_;
```
Comment (JiayuZzz) [inline engine/kv_engine.hpp:76]: maybe vesited_skiplist_in_recovery? and no "_" in the end for public memble.

File: engine/kv_engine.cpp [hunk 5]
```diff
@@ -380,9 +385,22 @@ Status KVEngine::RestoreSortedRecord(DLDataEntry *pmem_data_entry,
         return Status::MemoryOverflow;
       }
       new_hash_offset = (uint64_t)dram_node;
+// RESTORE_SKIPLIST_STRIDE: Select a data entry every 10000 into restored
+// skiplist map for multi-thread restoring large skiplist.
+#define RESTORE_SKIPLIST_STRIDE 10000
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:390]: Don't define here, add a "constexpr uint64_t kRestoreSkiplistStride = 10000;" at begin of the source file.

File: tests/tests.cpp [hunk 6]
```diff
@@ -746,6 +750,61 @@ TEST_F(EngineBasicTest, TestSortedHotspot) {
   delete engine;
 }
 
+TEST_F(EngineBasicTest, TestMultiThreadSortedRestore) {
+  int num_threads = 48;
+  int num_collections = 16;
+  configs.max_write_threads = num_threads;
+  configs.opt_restore_sorted = true;
+  configs.populate_pmem_space = false;
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply tests/tests.cpp:758]: I want to quickly run this test， would you say `populate_pmem_space`?

File: engine/skiplist.hpp [hunk 7]
```diff
@@ -266,4 +276,32 @@ class SortedIterator : public Iterator {
...
+  DealWithOtherHeight(uint64_t thread_id, SkiplistNode *cur_node, int heightm,
+                      const std::shared_ptr<PMEMAllocator> &pmem_allocator);
+
+  SkiplistNode *GetSortedOffset(int height);
+
+  void LinkedNode(uint64_t thread_id, int height, const KVEngine *engine);
+
+  void Rebuild(const KVEngine *engine);
+
+  static std::unordered_map<uint64_t, std::pair<bool, SkiplistNode *>>
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.hpp:297]: If limited one kvdk instance, please use  single instance; I think it is suitable to use static; Only a parallelrebuildsorted object;

File: engine/skiplist.cpp [hunk 8]
```diff
@@ -232,4 +282,249 @@ Skiplist::InsertDataEntry(Splice *insert_splice, DLDataEntry *inserting_entry,
...
+        (DLDataEntry *)engine->pmem_allocator_->offset2addr(it->first);
+    pmem::obj::string_view key = cur_data_entry->Key();
+
+    Status s = engine->hash_table_->Search(
+        engine->hash_table_->GetHint(key), key, SortedDataEntryType,
+        &hash_entry, &data_entry, &entry_base, HashTable::SearchPurpose::Read);
+    assert(s == Status::Ok || s == Status::NotFound);
+    if (s == Status::NotFound ||
+        hash_entry.header.offset_type == HashOffsetType::DLDataEntry) {
+      it = entries_offsets_.erase(it);
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.cpp:507]: no, the one iter;

File: engine/kv_engine.hpp (hunk 4 above)
Comment (sherry-1001) [reply engine/kv_engine.hpp:76]: ok

File: engine/skiplist.cpp [hunk 9]
```diff
@@ -232,4 +282,249 @@ Skiplist::InsertDataEntry(Splice *insert_splice, DLDataEntry *inserting_entry,
...
+      }
+    }
+  }
+  thread_cache_node_[thread_id].clear();
+}
+
+SkiplistNode *ParallelRebuildSorted::GetSortedOffset(int height) {
+  std::lock_guard<std::mutex> kv_mux(map_mu_);
+  for (auto &kv : entries_offsets_) {
+    if (!kv.second.first && kv.second.second->Height() >= height - 1) {
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.cpp:388]: good suggestions

File: tests/tests.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (JiayuZzz) [reply tests/tests.cpp:758]: yes

File: engine/skiplist.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (JiayuZzz) [reply engine/skiplist.hpp:297]: think about this:
std::string path1, path2;
Engine* engine1, engine2;
Engine::Open(path1, &engine1, ...);
Engine::Open(path2, &engine2, ...);

then this map will be used by both engines.

File: engine/skiplist.hpp [hunk 10]
```diff
@@ -266,4 +276,32 @@ class SortedIterator : public Iterator {
   std::shared_ptr<PMEMAllocator> pmem_allocator_;
   DLDataEntry *current;
 };
+
+class KVEngine;
+class ParallelRebuildSorted {
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.hpp:281]: sounds good 😊

[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.hpp:281]: it is `rebuilder`, so the rebuild of a skiplist should be included.

File: include/kvdk/configs.hpp [hunk 11]
```diff
@@ -82,6 +82,11 @@ struct Configs {
 
   // Log information to show
   LogLevel log_level = LogLevel::INFO;
+
+  // Multi-thread recovery a skiplist. The optimization can get better
+  // performance when having few large skiplists. Default is to close
+  // optimization.
+  bool opt_restore_sorted = false;
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply include/kvdk/configs.hpp:89]: I used `restore_large_sorted_collection`  to express this setting config  optimize for large skiplist

File: engine/skiplist.hpp [hunk 12]
```diff
@@ -266,4 +276,32 @@ class SortedIterator : public Iterator {
...
+
+  void
+  DealWithOtherHeight(uint64_t thread_id, SkiplistNode *cur_node, int heightm,
+                      const std::shared_ptr<PMEMAllocator> &pmem_allocator);
+
+  SkiplistNode *GetSortedOffset(int height);
+
+  void LinkedNode(uint64_t thread_id, int height, const KVEngine *engine);
+
+  void Rebuild(const KVEngine *engine);
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/skiplist.hpp:295]: I don't want to `SortedCollectionRebuilder` and `kvEngine`   become members of each other, it doesn’t fit the design principle very well;

**Review by JiayuZzz — APPROVED** [review]
LGTM

**Pull request merged**